go-controller/pkg/node/gateway_init_linux_test.go

```diff
@@ -298,13 +299,169 @@ cookie=0x0, duration=8366.597s, table=1, n_packets=10641, n_bytes=10370087, prio
+			"ovs-vsctl --timeout=15 set bridge br-local other-config:hwaddr=" + brLocalnetMAC,
+		})
+		fexec.AddFakeCmd(&ovntest.ExpectedCmd{
+			Cmd:    "ovs-vsctl --timeout=15 --if-exists get Open_vSwitch . external_ids:ovn-bridge-mappings",
+			Output: "",
+		})
+		fexec.AddFakeCmdsNoOutputNoError([]string{
+			"ovs-vsctl --timeout=15 set Open_vSwitch . external_ids:ovn-bridge-mappings=" + util.PhysicalNetworkName + ":br-local",
+			"ovs-vsctl --timeout=15 --if-exists del-port br-local " + legacyLocalnetGatewayNextHopPort +
+				" -- --may-exist add-port br-local " + localnetGatewayNextHopPort + " -- set interface " + localnetGatewayNextHopPort + " type=internal mtu_request=" + mtu + " mac=00\\:00\\:a9\\:fe\\:21\\:01",
+		})
+		fexec.AddFakeCmd(&ovntest.ExpectedCmd{
+			Cmd:    "ovs-vsctl --timeout=15 --if-exists get Open_vSwitch . external_ids:system-id",
+			Output: systemID,
+		})
+		fexec.AddFakeCmd(&ovntest.ExpectedCmd{
+			Cmd: "ip rule",
+			Output: "0:	from all lookup local\n32766:	from all lookup main\n32767:	from all lookup default\n",
+		})
+		fexec.AddFakeCmdsNoOutputNoError([]string{
+			"ip rule add from all table " + localnetGatewayExternalIDTable,
+		})
+		fexec.AddFakeCmdsNoOutputNoError([]string{
+			"ip route list table " + localnetGatewayExternalIDTable,
+		})
+
+		existingNode := v1.Node{ObjectMeta: metav1.ObjectMeta{
+			Name: nodeName,
+		}}
+
+		fakeOvnNode.start(ctx,
+			&v1.NodeList{
+				Items: []v1.Node{
+					existingNode,
+				},
+			},
+		)
+
+		nodeAnnotator := kube.NewNodeAnnotator(&kube.Kube{fakeOvnNode.fakeClient}, &existingNode)
+		err := util.SetNodeHostSubnetAnnotation(nodeAnnotator, subnets)
+		Expect(err).NotTo(HaveOccurred())
+		err = nodeAnnotator.Run()
+		Expect(err).NotTo(HaveOccurred())
+
+		err = testNS.Do(func(ns.NetNS) error {
+			defer GinkgoRecover()
+
+			config.IPv4Mode = false
+			config.IPv6Mode = false
+			for _, subnet := range subnets {
+				if utilnet.IsIPv6CIDR(subnet) {
+					config.IPv6Mode = true
+				} else {
+					config.IPv4Mode = true
+				}
+			}
+
+			err = fakeOvnNode.node.initLocalnetGateway(subnets, nodeAnnotator)
+			Expect(err).NotTo(HaveOccurred())
+			// Check if IP has been assigned to LocalnetGatewayNextHopPort
+			link, err := netlink.LinkByName(localnetGatewayNextHopPort)
+			Expect(err).NotTo(HaveOccurred())
+			addrs, err := netlink.AddrList(link, 0)
+			Expect(err).NotTo(HaveOccurred())
+
+			var foundAddr bool
+			for _, expectedAddr := range brNextHopCIDRs {
+				foundAddr = false
+				for _, a := range addrs {
+					if a.IP.Equal(expectedAddr.IP) && bytes.Equal(a.Mask, expectedAddr.Mask) {
+						foundAddr = true
+						break
+					}
+				}
+				if !foundAddr {
+					break
+				}
+			}
+			Expect(foundAddr).To(BeTrue())
+			return nil
+		})
+		Expect(err).NotTo(HaveOccurred())
+
+		Expect(fexec.CalledMatchesExpected()).To(BeTrue(), fexec.ErrorDesc)
+
+		for i := 0; i < len(ipts); i++ {
+			Expect(ipts[i].MatchState(expectedIpTableRules[i])).NotTo(HaveOccurred())
```

Comment:
This breaks the pseudo-English of ginkgo/gomega and makes it look like the test is asserting that a match didn't occur. You have to have an actual error variable for it to be clear:

    err = ipts[i].MatchState(expectedIpTableRules[i])
    Expect(err).NotTo(HaveOccurred())

Reply:
Done